monitor/server.cpp

```diff
@@ -20,146 +20,210 @@
+	m_acceptor.async_accept(*socket, [socket, this](const boost::system::error_code &err) {
+		if (!err) {
+			std::make_shared<handler>(m_monitor, std::move(*socket))->start();
+		}
 
-	async_accept();
+		async_accept();
+	});
 }
 
-
 void server::stop() {
 	m_io_service.stop();
 }
 
 void handler::async_read() {
 	auto self(shared_from_this());
-	m_socket.async_read_some(boost::asio::buffer(m_buffer),
-	                         std::bind(&handler::handle_read, self,
-	                                   std::placeholders::_1,
-	                                   std::placeholders::_2));
-}
 
-void handler::handle_read(const boost::system::error_code &err, size_t size) {
-	if (err) {
-		close();
-		return;
-	}
+	m_start_ts = std::chrono::system_clock::now();
+	m_socket.async_read_some(boost::asio::buffer(m_buffer), [self, this](const boost::system::error_code &err,
+	                                                                     size_t size) {
+		if (err) {
+			close();
+			return;
+		}
 
-	auto req = parse_request(size);
-	std::string content = "";
+		m_recv_ts = std::chrono::system_clock::now();
 
-	if (req > 0) {
-		DNET_LOG_DEBUG(m_monitor.node(),
-		               "monitor: server: got statistics request for categories: {:x} from: {}:{}", req,
-		               m_remote, m_socket.remote_endpoint().port());
-		content = m_monitor.get_statistics().report(req);
-	}
+		const auto req = parse_request(size);
+		std::string content = "";
+
+		if (req > 0) {
+			DNET_LOG_DEBUG(m_monitor.node(),
+			               "monitor: http-server: got statistics request for categories: {:x} from: {}",
+			               req, m_remote);
+			content = m_monitor.get_statistics().report(req);
+		}
+
+		m_response = make_reply(req, std::move(content));
 
-	std::string reply = make_reply(req, content);
-	async_write(reply);
+		m_collect_ts = std::chrono::system_clock::now();
+
+		async_write();
+
+	});
 }
 
-void handler::async_write(std::string data) {
+void handler::async_write() {
+	DNET_LOG_DEBUG(m_monitor.node(), "monitor: http-server: send requested statistics: started: {}, size: {}",
+	               m_remote, m_response.size());
+
 	auto self(shared_from_this());
-	m_report = std::move(data);
-	DNET_LOG_DEBUG(m_monitor.node(), "monitor: server: send requested statistics: started: {}:{}, size: {}",
-	               m_remote, m_socket.remote_endpoint().port(), m_report.size());
-	boost::asio::async_write(m_socket, boost::asio::buffer(m_report),
-	                         std::bind(&handler::handle_write, self));
-}
 
-void handler::handle_write() {
-	DNET_LOG_DEBUG(m_monitor.node(), "monitor: server: send requested statistics: finished: {}:{}", m_remote,
-	               m_socket.remote_endpoint().port());
-	close();
+	boost::asio::async_write(m_socket, boost::asio::buffer(m_response), [self, this](const boost::system::error_code &/*err*/,
+	                                                                         const long unsigned int &size) {
```

Comment:
No, it isn't. `boost::asio::async_write` calls callback when whole buffer is sent or error is occurred.